File: slang-checks/src/main/java/org/sonarsource/slang/checks/UnusedPrivateMethodCheck.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.checks;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.sonar.check.Rule;
import org.sonarsource.slang.api.ClassDeclarationTree;
import org.sonarsource.slang.api.FunctionDeclarationTree;
import org.sonarsource.slang.api.IdentifierTree;
import org.sonarsource.slang.api.Tree;
import org.sonarsource.slang.checks.api.CheckContext;
import org.sonarsource.slang.checks.api.InitContext;
import org.sonarsource.slang.checks.api.SlangCheck;
import org.sonarsource.slang.checks.utils.FunctionUtils;
import org.sonarsource.slang.utils.SyntacticEquivalence;

import static org.sonarsource.slang.utils.SyntacticEquivalence.getUniqueIdentifier;

@Rule(key = "S1144")
public class UnusedPrivateMethodCheck implements SlangCheck {

  @Override
  public void initialize(InitContext init) {
    init.register(ClassDeclarationTree.class, this::processClassDeclaration);
  }

  protected void processClassDeclaration(CheckContext context, ClassDeclarationTree classDeclarationTree) {
    // only verify the outermost class in the file, to avoid raising the same issue multiple times
    if (context.ancestors().stream().noneMatch(ClassDeclarationTree.class::isInstance)) {
      reportUnusedPrivateMethods(context, classDeclarationTree);
    }
  }

  protected void reportUnusedPrivateMethods(CheckContext context, ClassDeclarationTree classDeclarationTree) {
    MethodAndIdentifierCollector methodAndIdentifierCollector = new MethodAndIdentifierCollector(classDeclarationTree.descendants());
    methodAndIdentifierCollector.getMethodDeclarations().stream()
      .filter(this::isValidPrivateMethod)
      .forEach(tree -> {
        IdentifierTree identifier = tree.name();
        if (identifier != null && isUnusedMethod(identifier, methodAndIdentifierCollector.getUsedUniqueIdentifiers())) {
          String message = String.format("Remove this unused private \"%s\" method.", identifier.name());
          context.reportIssue(tree.rangeToHighlight(), message);
        }
      });
  }

  protected boolean isValidPrivateMethod(FunctionDeclarationTree method) {
    return FunctionUtils.isPrivateMethod(method) && !FunctionUtils.isOverrideMethod(method);
  }

  protected boolean isUnusedMethod(IdentifierTree identifier, Set<String> usedIdentifierNames) {
    return !usedIdentifierNames.contains(getUniqueIdentifier(identifier));
  }

  protected static class MethodAndIdentifierCollector {
    private Set<FunctionDeclarationTree> methodDeclarations = new HashSet<>();
    private Set<String> usedUniqueIdentifiers;

    Set<FunctionDeclarationTree> getMethodDeclarations() {
      return methodDeclarations;
    }
    public Set<String> getUsedUniqueIdentifiers() {
      return usedUniqueIdentifiers;
    }

    public MethodAndIdentifierCollector(Stream<Tree> descendants) {
      Set<IdentifierTree> usedIdentifiers = new HashSet<>();
      descendants.forEach(tree -> {
        if (tree instanceof FunctionDeclarationTree && !((FunctionDeclarationTree)tree).isConstructor()) {
          methodDeclarations.add(((FunctionDeclarationTree) tree));
        } else if (tree instanceof IdentifierTree) {
          usedIdentifiers.add((IdentifierTree) tree);
        }
      });

      usedIdentifiers.removeAll(methodDeclarations.stream()
        .map(FunctionDeclarationTree::name)
        .collect(Collectors.toSet()));

      usedUniqueIdentifiers = usedIdentifiers.stream()
        .filter(Objects::nonNull)
        .map(SyntacticEquivalence::getUniqueIdentifier)
        .collect(Collectors.toCollection(HashSet::new));
    }

  }
}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnusedPrivateMethodCheck.kt
/*
 * SonarSource Kotlin
 * Copyright (C) 2018-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.checks

import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtClass
import org.jetbrains.kotlin.psi.KtNameReferenceExpression
import org.jetbrains.kotlin.psi.KtNamedFunction
import org.jetbrains.kotlin.psi.KtOperationReferenceExpression
import org.jetbrains.kotlin.psi.psiUtil.anyDescendantOfType
import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
import org.jetbrains.kotlin.psi.psiUtil.isPrivate
import org.jetbrains.kotlin.resolve.BindingContext
import org.sonar.check.Rule
import org.sonarsource.kotlin.api.AbstractCheck
import org.sonarsource.kotlin.api.isInfix
import org.sonarsource.kotlin.plugin.KotlinFileContext

// Serializable method should not raise any issue in Kotlin.
private val IGNORED_METHODS: Set<String> = setOf(
    "writeObject",
    "readObject",
    "writeReplace",
    "readResolve",
    "readObjectNoData",
)

private val COMMON_ANNOTATIONS = listOf(
    "kotlin.OptIn", 
    "kotlin.Suppress",
    "kotlinx.coroutines.DelicateCoroutinesApi",
    "kotlin.jvm.Throws",
)

@Rule(key = "S1144")
class UnusedPrivateMethodCheck : AbstractCheck() {

    override fun visitClass(klass: KtClass, context: KotlinFileContext) {
        if (!klass.isTopLevel()) return
        klass.collectDescendantsOfType<KtNamedFunction> { it.shouldCheckForUsage(context.bindingContext) }
            .forEach {
                val functionName = it.name!!
                if (!IGNORED_METHODS.contains(functionName) && !it.isReferencedIn(klass, functionName)) {
                    // Anonymous functions can't be private, so nameIdentifier is always present
                    context.reportIssue(it.nameIdentifier!!, """Remove this unused private "$functionName" method.""")
                }
            }
    }

    private fun KtNamedFunction.isReferencedIn(klass: KtClass, name: String) =
        klass.hasReferences(name) || (isInfix() && klass.hasInfixReferences(name))

    private fun KtClass.hasReferences(name: String) =
        anyDescendantOfType<KtNameReferenceExpression> { it.getReferencedName() == name }

    private fun KtClass.hasInfixReferences(name: String) =
        anyDescendantOfType<KtOperationReferenceExpression> { it.getReferencedName() == name }

    private fun KtNamedFunction.shouldCheckForUsage(bindingContext: BindingContext) =
        isPrivate()
            && !hasModifier(KtTokens.OPERATOR_KEYWORD)
            && (annotationEntries.isEmpty() || annotatedWithCommonAnnotations(bindingContext))

    private fun KtNamedFunction.annotatedWithCommonAnnotations(bindingContext: BindingContext) =
        bindingContext.get(BindingContext.FUNCTION, this)
            ?.annotations
            ?.all { it.fqName?.asString() in COMMON_ANNOTATIONS }
            ?: false

}
